README.adoc

```diff
@@ -1026,6 +1026,24 @@ end
 While enforcing table defaults only in Rails is suggested by many Rails developers, it's an extremely brittle approach that leaves your data vulnerable to many application bugs.
 And you'll have to consider the fact that most non-trivial apps share a database with other applications, so imposing data integrity from the Rails app is impossible.
 
+=== 3-state Boolean [[three-state-boolean]]
+
+With SQL databases, if a boolean column is not given a default value, it ends up having three possible values: `true`, `false` and `NULL`.
+Boolean operators [work in unexpected ways](https://en.wikipedia.org/wiki/Three-valued_logic) with `NULL`.
+
+E.g. in SQL queries, `true AND NULL` is `NULL` (not false), `true AND NULL OR false` is `NULL` (not false). This can make SQL queries return unexpected results.
+
+To avoid such situations, boolean columns should always have a default value and `NOT NULL` constraint.
+
+[source,ruby]
+----
+# bad - boolean without default
+add_column :users, :active, :boolean
+
+# good - boolean with default and restricted with null
```

Comment:
Sorry for nitpicking, but I'm not too confident if "restricted with null" explains what `null: false` does.
I think `with restricted NULL`. A quick search over https://guides.rubyonrails.org/active_record_migrations.html didn't yield any better wording, unfortunately.

Reply:
changed